File: python_solutions/q031.py
# In the United Kingdom the currency is made up of pound (£) and pence (p). There are eight coins in general circulation:

# 1p, 2p, 5p, 10p, 20p, 50p, £1 (100p), and £2 (200p).
# It is possible to make £2 in the following way:

# 1×£1 + 1×50p + 2×20p + 1×5p + 1×2p + 3×1p
# How many different ways can £2 be made using any number of coins?

# intuition:
# top down dp


def compute():
    target = 200
    coins = [1, 2, 5, 10, 20, 50, 100, 200]
    dp = [1] + [0] * target
    
    for i in range(len(coins)):
        for j in range(coins[i], target + 1):
            dp[j] += dp[j - coins[i]]
    print(dp)
    return dp[target]


if __name__ == "__main__":
    print(compute())

File: java_solutions/q031.java
package java_solutions;

public class q031 {
  public static void main(String[] args) {
    System.out.println(new q031().compute());
  }

  public int compute() {
    int target = 200;
    int[] coins = {1, 2, 5, 10, 20, 50, 100, 200};
    int[] dp = new int[target + 1];
    dp[0] = 1;

    for (int i = 0; i < coins.length; i++) {
      for (int j = coins[i]; j <= target; j++) {
        dp[j] += dp[j - coins[i]];
      }
    }

    return dp[target];
  }
}
